### ISSUE TITLE
Store temporary files in memory

### ISSUE THREAD

**Issue by kenkendk** [duplicati/duplicati#2816]
As the temporary files are usually small (say 50-200MB) they can easily fit into memory on most machines (small ARM devices like RPI or Synology excluded).

Experiments with setting a RAM-disk as the temporary folder shows significant speedup and will reduce write-wear on SSD disks.

This issue suggests that the temporary files for `dblock`, `dindex` and `dlist` are stored in memory, if the `--volume-size` is less than 10% of the memory. This should of course be configurable, for users who want to control what happens.

This change requires that the `ICompressionModule` interface is changed slightly such that it accepts a `System.IO.Stream` as input instead of the current path-`string`.

The `BackendManager` needs to be changed a bit as it relies on the files being present. and it should also support streams.

An optional, and more extensive solution, would be to replace calls to `System.IO.File.Open` that work on temporary files, such that a special kind of path, like: `tempfile://xyz` can be detected, and redirect to some internal storage dictionary with temporary files being mapped to (memory-)streams.

**Comment by TekkiWuff** [comment]
@kenkendk

I'm currently tinkering on this one a bit. Nothing too serious yet but nevertheless I have a question regarding the encryption and decryption.

There are multiple places in the code that read data from a temporary file, de-/encrypt the data and write it back to a new temporary file. Notably:
 - https://github.com/duplicati/duplicati/blob/28c6effac2dde22ee966f9b72b751d476ab3ccfd/Duplicati/Library/Main/BackendManager.cs#L182
 - https://github.com/duplicati/duplicati/blob/28c6effac2dde22ee966f9b72b751d476ab3ccfd/Duplicati/Library/Main/BackendManager.cs#L950
 
 If I want to store the tempory file data in memory, then it seems like the usage of temporary files should be removed here as well.
 Now I wonder why the data is not simply passed through the encryption layer via Streams. The `IEncryption` interface seems to have the nessecary `Stream Encrypt(Stream input)` and `Stream Decrypt(Stream input)` methods that both the `AESEncryption` and `GPGEncryption` classes implement with actual code.
 
The only hint I found, is a comment (see second link) that mentions something about not being able to differentiate in which part of the stack an exception resides. Why is this an issue and is this even still relevant?

**Comment by kenkendk** [comment]
The reason for not using a stream, is that I have stumbled on different cases where this fails. It is my understanding that there is something in the .Net Crypto API (padding?) that does not work with non-seekable streams.

The part with the exception is a temporary fix to do some encryption in parallel. For the `concurrent_processing` branch which I am working on, the compressed volume is just encrypted by itself. This code will go away, so no need to handle that part.

**Comment by BMurri** [comment]
No reason we can't have a fully seekable in-memory stream

**Comment by MrElliwood** [comment]
There's now a bounty of $40

**Comment by Jojo-1000** [comment]
Most of the changes originally mentioned are implemented by now.

Proof of concept implementation: [Jojo-1000:mem-temporary-files](https://github.com/Jojo-1000/duplicati/tree/mem-temporary-files)

- Interface to allow temporary files on disk or in memory
- Some backends require filenames (rsync), convert to actual file by writing to disk in that case or if `--disable-streaming-transfers` is set

Missing:
- Detection of available memory
- Spillover to disk if memory becomes too full
- Configurable threshold
- Check whether the backend allows it when creating the temp file instead of when reading

The speed is not significantly higher compared to NVME SSD (probably due to RAM caching on the SSD). After discussion with @gpatel-fr, it was decided to postpone work on this for now.
